Objects/genobject.c

```diff
@@ -1797,16 +1797,22 @@ async_gen_athrow_send(PyAsyncGenAThrow *o, PyObject *arg)
     PyFrameObject *f = gen->gi_frame;
     PyObject *retval;
 
-    if (f == NULL || f->f_stacktop == NULL ||
-            o->agt_state == AWAITABLE_STATE_CLOSED) {
+    if (o->agt_state == AWAITABLE_STATE_CLOSED) {
         PyErr_SetString(
             PyExc_RuntimeError,
             "cannot reuse already awaited aclose()/athrow()");
         return NULL;
     }
 
+    if (f == NULL || f->f_stacktop == NULL) {
+        o->agt_state = AWAITABLE_STATE_CLOSED;
+        PyErr_SetNone(PyExc_StopIteration);
+        return NULL;
+    }
+
```

Comment:
These two hunks are the main change. Originally, this code raise `StopIteration` if the generator had no frame. The fix for bpo-39386 added the check for the athrow coroutine being in `AWAITABLE_STATE_CLOSED` to the same `if` statement, and switched it to raise `RuntimeError` in both cases. But in fact, the generator having no frame *should* raise `StopIteration`; it's only the `AWAITABLE_STATE_CLOSED` case that should raise `RuntimeError`. So I split the `AWAITABLE_STATE_CLOSED` logic out into its own block, and restored the old `if (f == NULL || f->f_stacktop)` logic.

Everything else is just going through to make sure that we're setting `AWAITABLE_STATE_CLOSED` at the right times.